TYPO3.Flow/Classes/TYPO3/Flow/Aop/Builder/ProxyClassBuilder.php

```diff
@@ -540,8 +540,21 @@ protected function addBuildMethodsAndAdvicesCodeToClass($className, ClassNameInd
         }
 
         $callBuildMethodsAndAdvicesArrayCode = "        if (method_exists(get_parent_class(), 'Flow_Aop_Proxy_buildMethodsAndAdvicesArray') && is_callable('parent::Flow_Aop_Proxy_buildMethodsAndAdvicesArray')) parent::Flow_Aop_Proxy_buildMethodsAndAdvicesArray();\n";
-        $proxyClass->getConstructor()->addPreParentCallCode($callBuildMethodsAndAdvicesArrayCode);
-        $proxyClass->getMethod('__wakeup')->addPreParentCallCode($callBuildMethodsAndAdvicesArrayCode);
+
+        $proxyMethods = ['__construct' => null, '__wakeup' => null];
+        if ($this->reflectionService->isMethodFinal($className, '__construct') === false) {
+            $proxyClass->getConstructor()->addPreParentCallCode($callBuildMethodsAndAdvicesArrayCode);
+            $proxyMethods['__construct'] = true;
+        }
+
+        if ($this->reflectionService->isMethodFinal($className, '__wakeup') === false) {
+            $proxyClass->getMethod('__wakeup')->addPreParentCallCode($callBuildMethodsAndAdvicesArrayCode);
+            $proxyMethods['__wakeup'] = true;
+        }
+
+        foreach (array_keys(array_filter($proxyMethods, 'is_null')) as $methodName) {
```

Comment:
Wouldn't your implementation log two messages exactly the same way??
Anyways it can ofc be changed to sth like
```php
$skippedMethods = [];
if ($this->reflectionService->isMethodFinal($className, '__construct')) {
    $skippedMethods[] = '__construct';
} else {
    $proxyClass->getConstructor()->addPreParentCallCode($callBuildMethodsAndAdvicesArrayCode);
}

if ($this->reflectionService->isMethodFinal($className, '__wakeup')) {
    $skippedMethods[] = '__wakeup';
} else {
    $proxyClass->getMethod('__wakeup')->addPreParentCallCode($callBuildMethodsAndAdvicesArrayCode);
}

if ($skippedMethods !== []) {
    $this->systemLogger->log(sprintf('Could not proxy the method(s) "%s" of class "%s", as it is declared final. AOP might not be correctly applied.', explode(', ', $skippedMethods), $className), LOG_DEBUG);
}
```
too.. Anyways, not a biggy of course.